include/swift/AST/LocalizationFormat.h

```diff
@@ -43,7 +131,29 @@ class LocalizationProducer {
 class YAMLLocalizationProducer final : public LocalizationProducer {
 public:
   std::vector<std::string> diagnostics;
-  explicit YAMLLocalizationProducer(std::string locale, std::string path);
+  explicit YAMLLocalizationProducer(llvm::StringRef filePath);
+  llvm::StringRef getMessageOr(swift::DiagID id,
+                               llvm::StringRef defaultMessage) const override;
+};
+
+class SerializedLocalizationProducer final : public LocalizationProducer {
+  using SerializedLocalizationTable =
+      llvm::OnDiskIterableChainedHashTable<LocalizationLookupInfo>;
+  llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> Buffer;
+  llvm::BitstreamCursor Cursor;
+  std::unique_ptr<SerializedLocalizationTable> SerializedTable;
+
+public:
+  explicit SerializedLocalizationProducer(llvm::StringRef filePath)
+      : Buffer(llvm::MemoryBuffer::getFileOrSTDIN(filePath)),
```

Comment:
Since we need to verify that `getFileOrSTDIN` returns a valid buffer, I'd suggest we pass `std::unique_ptr<MemoryBuffer>` as a constructor argument so caller handle opening of the buffer and producing a warning if needed.